docs(upgrade): expand upon change detection implications for downgraded…

agale123 opened the pull request:
… components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [v] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [v] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

agale123 committed:
docs(upgrade): expand upon change detection implications for downgrade components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

refactor(compiler-cli): implement `BabelAstFactory` and `AstHost`s (#38866)

This commit adds the `AstHost` interface, along with implementations for
both Babel and TS.

It also implements the Babel vesion of the `AstFactory` interface, along
with a linker specific implementation of the `ImportGenerator` interface.

These classes will be used by the new "ng-linker" to transform prelinked
library code using a Babel plugin.

PR Close #38866

refactor(docs-infra): more closely align `tslint.json` with CLI (#39018)

This commit re-organizes the `tslint.json` configuration file to more
closely align with the one generated by the latest Angular CLI for new
apps. This makes it easier to see the difference with new CLI versions
in the future and keep our `tslint.json` up-to-date (while keeping our
own rules).

NOTE:
This commit only re-orders rules and removes redundant ones. It does not
change the linting behavior.

PR Close #39018

build(docs-infra): enable the `no-redundant-jsdoc` tslint rule to align with CLI (#39018)

This commit enables the `no-redundant-jsdoc` tslint rule to more closely
align `tslint.json` with the one generated by the latest Angular CLI for
new apps.

PR Close #39018

build(docs-infra): remove the `no-string-literal` tslint rule to align with CLI (#39018)

This commit removes the `no-string-literal: false` tslint rule to
more closely align `tslint.json` with the one generated by the latest
Angular CLI for new apps.

PR Close #39018

build(docs-infra): update the `object-literal-key-quotes` tslint rule to align with CLI (#39018)

This commit updates the `object-literal-key-quotes` tslint rule to more
closely align `tslint.json` with the one generated by the latest Angular
CLI for new apps.

PR Close #39018

build(docs-infra): remove the `only-arrow-functions` tslint rule to align with CLI (#39018)

This commit removes the `only-arrow-functions: false` tslint rule to
more closely align `tslint.json` with the one generated by the latest
Angular CLI for new apps.

PR Close #39018

refactor(docs-infra): leave TODO comments for aligning tslint with CLI (#39018)

This commit adds some TODO comments in `tslint.json` regardling rules
that need to be enabled or removed to more closely align `tslint.json`
with the one generated by the latest Angular CLI for new apps.

Updating these rules generates a lot of linting failures, so fixing
them is outside the scope of this PR.

PR Close #39018

fix(core): use single quotes for relative link resolution migration to align with style guide (#39070)

This updates the migration to align with the style guide and work with default lint rules. It avoids a lint error on
newly migrated projects and fixes a test in the CLI repo.

PR Close #39070

Revert "fix(core): use single quotes for relative link resolution migration to align with style guide (#39070)" (#39082)

This reverts commit 889470663d2bcb82e8018996c7a751635491ebe7.

PR Close #39082

build(docs-infra): upgrade cli command docs sources to 65fb6788d (#39078)

Updating [angular#master](https://github.com/angular/angular/tree/master) from
[cli-builds#master](https://github.com/angular/cli-builds/tree/master).

Relevant changes in
[commit range](https://github.com/angular/cli-builds/compare/e526da605...65fb6788d):

**Modified**
- help/generate.json
- help/new.json

PR Close #39078

feat(dev-infra): add a command to verify NgBot YAML config syntax (#39071)

This commit adds a new command to the `ng-dev` suite, which verifies that the NgBot YAML config is
correct. It also adds this command to the `lint` CircleCI job so that we execute this check while
running CI.

This should help prevent syntax errors similar to the one introduced in:
https://github.com/angular/angular/commit/393ce5574b76ad1344554a6aa9e2e8a6c3591ac9

PR Close #39071

fix(dev-infra): correct matching logic for g3 comparison (#39084)

Corrects the matching logic for g3 comparison check, previously more matches were
found than were correct.

PR Close #39084

ci: update g3 synced file list (#39084)

Update the list of files synced into g3

PR Close #39084

build: upgrade angular build, integration/bazel and @angular/bazel package to rule_nodejs 2.2.0 (#37727)

Updates to rules_nodejs 2.2.0. This is the first major release in 7 months and includes a number of features as well
as breaking changes.

Release notes: https://github.com/bazelbuild/rules_nodejs/releases/tag/2.0.0

Features of note for angular/angular:

* stdout/stderr/exit code capture; this could be potentially be useful

* TypeScript (ts_project); a simpler tsc rule that ts_library that can be used in the repo where ts_library is too
  heavy weight

Breaking changes of note for angular/angular:

* loading custom rules from npm packages: `ts_library` is no longer loaded from `@npm_bazel_typescript//:index.bzl`
  (which no longer exists) but is now loaded from `@npm//@bazel/typescript:index.bzl`

* with the loading changes above, `load("@npm//:install_bazel_dependencies.bzl", "install_bazel_dependencies")` is
  no longer needed in the WORKSPACE which also means that yarn_install does not need to run unless building/testing
  a target that depends on @npm. In angular/angular this is a minor improvement as almost everything depends on @npm.

* @angular/bazel package is also updated in this PR to support the new load location; Angular + Bazel users that
  require it for ng_package (ng_module is no longer needed in OSS with Angular 10) will need to load from
  `@npm//@angular/bazel:index.bzl`. I investigated if it was possible to maintain backward compatability for the old
  load location `@npm_angular_bazel` but it is not since the package itself needs to be updated to load from
  `@npm//@bazel/typescript:index.bzl` instead of `@npm_bazel_typescript//:index.bzl` as it depends on ts_library
  internals for ng_module.

* runfiles.resolve will now throw instead of returning undefined to match behavior of node require

Other changes in angular/angular:

* integration/bazel has been updated to use both ng_module and ts_libary with use_angular_plugin=true.
  The latter is the recommended way for rules_nodejs users to compile Angular 10 with Ivy. Bazel + Angular ViewEngine is
  supported with @angular/bazel <= 9.0.5 and Angular <= 8. There is still Angular ViewEngine example on rules_nodejs
  https://github.com/bazelbuild/rules_nodejs/tree/stable/examples/angular_view_engine on these older versions but users
  that want to update to Angular 10 and are on Bazel must switch to Ivy and at that point ts_library with
  use_angular_plugin=true is more performant that ng_module. Angular example in rules_nodejs is configured this way
  as well: https://github.com/bazelbuild/rules_nodejs/tree/stable/examples/angular. As an aside, we also have an
  example of building Angular 10 with architect() rule directly instead of using ts_library with angular plugin:
  https://github.com/bazelbuild/rules_nodejs/tree/stable/examples/angular_bazel_architect.

NB: ng_module is still required for angular/angular repository as it still builds ViewEngine & @angular/bazel
also provides the ng_package rule. ng_module can be removed in the future if ViewEngine is no longer needed in
angular repo.

* JSModuleInfo provider added to ng_module. this is for forward compat for future rules_nodejs versions.
  @josephperrott, this touches `packages/bazel/src/external.bzl` which will make the sync to g3 non-trivial.

PR Close #37727

docs(upgrade): expand upon change detection implications for downgrade components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for downgrade components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

agale123 committed:
docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

agale123 committed:
docs(upgrade): expand upon change detection implications for downgrade components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

agale123 committed:
docs(upgrade): expand upon change detection implications for downgrade components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

agale123 committed:
docs(upgrade): expand upon change detection implications for downgrade components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in upgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

agale123 committed:
docs(upgrade): expand upon change detection implications for upgraded components

From reading this guide it isn't clear that every AngularJS digest cycle triggers change detection in downgraded components. This change makes that clear and details how to avoid that pitfall which can lead to poor performance when using a hybrid app. I could also move this to be a tip when downgradeComponent is first called.

Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -362,7 +362,10 @@ the component's inputs will be watched using AngularJS change detection.
 are set. You can also hook into the changes by implementing the
 [OnChanges](api/core/OnChanges) interface in the component,
-just like you could if it hadn't been downgraded.
+just like you could if it hadn't been downgraded. By default, change
+detection will also run for every AngularJS `$digest` cycle. If you wish
+to only have change detection run when the inputs change, set
+`propagateDigest: false` when downgrading components.
```
Nice addition!

As you mentioned in the commit message, since this is a high-level overview, I think it would be better to move it as a tip in the [Using Angular Components from Angularjs Code][1] section.

##
Also, I would suggest a couple of minor changes:
- Make it more clear that we are talking about change-detecting the specific component (not the whole Angular app).
- link to the `downgradeComponent()` API docs, where people can find more info about how to set `propagateDigest`.

I.e. something like:

```
By default, Angular change detection will also run on the component for every AngularJS `$digest` cycle.
If you wish to only have change detection run when the inputs change, you can set `propagateDigest` to `false` when [downgrading components](api/upgrade/static/downgradeComponent).
```

[1]: https://angular.io/guide/upgrade#using-angular-components-from-angularjs-code

agale123 replied:
Thanks for the suggestions! From elsewhere in the doc, it looks like if I just put downgradeComponent in back ticks then it links to the correct docs. 

And I definitely flipped upgrade/downgrade in the title.

gkalpak reviewed (changes requested):
Thx, @agale123! I've left a couple fo suggestions in comments.

BTW, I think you meant "downgraded components" (instead of "upgraded components") in the commit message :wink:

agale123 committed:
Apply suggested fixes.

Moves the new text describing propagateDigest to false to the "Using Angular Components from Angularjs Code" section.

gkalpak commented on aio/content/guide/upgrade.md:
```diff
@@ -479,6 +479,15 @@ using the `downgradeComponent()` method. The result is an AngularJS
+By default, Angular change detection will also run on the component for every
+AngularJS `$digest` cycle. If you wish to only have change detection run when
+the inputs change, you can set `propagateDigest` to `false` when calling
+`downgradeComponent`.
```
Using parenthesis after a function or metho name helps make the reader aware it is a callable function of method (and not something else). Consider chaning this to:

```suggestion
`downgradeComponent()`.
```

gkalpak reviewed (changes requested):
Thx for making the changes, @agale123 🚀
I left an optional suggestion, but otherwise the changes lgtm.

To make CI green, you need to squash all commits into one (and fix the commit message to say `downgraded` instead of `upgraded`) and rebase on the latest master.

agale123 committed:
Update aio/content/guide/upgrade.md

Co-authored-by: George Kalpakas <kalpakas.g@gmail.com>

agale123 committed:
Merge branch 'master' of https://github.com/agale123/angular

agale123 committed:
Merge branch 'master' of https://github.com/agale123/angular

agale123 committed:
Merge branch 'master' of https://github.com/agale123/angular

agale123 committed:
Merge branch 'master' of https://github.com/agale123/angular

agale123 committed:
Merge branch 'master' of https://github.com/agale123/angular

agale123 commented:
> Thx for making the changes, @agale123 🚀
> I left an optional suggestion, but otherwise the changes lgtm.
> 
> To make CI green, you need to squash all commits into one (and fix the commit message to say `downgraded` instead of `upgraded`) and rebase on the latest master.

I seem to have made a mess of my commits so I might just make a new PR with everything clean because I'm not really familiar with git and I'm not sure how to merge everything

agale123 commented:
Going to start from scratch because I made a mess of this commit

agale123 commented:
Created #39100 as a cleaner version since my commits got messy
